doc/man/man1/lmake.1
.TH LMAKE 1 "18 October 2024" "Doliam" "User Commands"
.SH NAME
lmake \- a reliable & scalable tool to execute a full flow

.SH SYNOPSIS
\fBlmake\fR [\fIOPTION\fR]... [\fITARGET\fR]...

.SH DESCRIPTION
.LP
The \fBlmake\fR utility determines automatically which pieces of a flow need to be executed and launch them.
This is based in the historical \fImake\fR program initially written by Stuart Feldman in 1976 with notable differences :
.IP \[bu]
\fBlmake\fR is reliable.
It ensures that necessary jobs are run in adequate order so that the asked files are in the same state as if all derived files were removed and all jobs were run.
In particular, this includes automatic discovery of actual dependencies.
.IP \[bu]
\fBlmake\fR is scalable.
Millions of files can be derived easily (job execution may require heavy resources, but \fBlmake\fR will stay light).
.IP \[bu]
The makefile is called \fILmakefile.py\fR and is written in Python3.
.IP \[bu]
Rule matching is based of regular expressions, not just a single wildcard (% with \fImake\fR).
.IP \[bu]
Recipies can be written in \fBbash\fR (1) or \fBpython\fR (1).
.IP \[bu]
Several \fBlmake\fR commands can be run simultaneously in the same directory.
Jobs needed by several of them will be shared.
Simlarly sources files can be edited without risks while \fBlmake\fR is running.
Derived files will be generated according to the state of the source files at the time the command was launched.
If this is not possible because an old version of a source file need to be accessed after it has been modified, an error will be generated.
Rerunning the same \fBlmake\fR command will take into account the new content of such source files.
.IP \[bu]
File states are based on content (using a checksum) rather than dates.
This makes it possible to avoid running a dependent job if a file has been reconstructed identically to its previous content.
.LP
\fBlmake\fR manages a fully coherent directory called a repository.
When it starts, it first determine the root of the repository (cf. \fBFILES\fR below).
.LP
To reach these goals, \fBlmake\fR maintains a state in the \fILMAKE\fR directory, instrument jobs (to catch all file accesses) during their execution
and launches a daemon (called \fBlmakeserver\fR) that can be shared between several concurrent invocations of \flmake\fR.
.LP
\fBlmake\fR generates an output line for each significant event :
.IP \[bu]
When a job starts if it is long (duration that qualifies a job as long is configurable). This start line is deemed of secondary importance.
.IP \[bu]
When a job terminates.  Such lines are followed by the content of stderr if any and status is known.
Typically the job status is not known at the end of execution when a new dep was discovered that turned out to be out-of-date.
In that case, the dep is rebuilt, and if the new content is identical to the old one, \fBlmake\fR can decide that the job was run with up-to-date deps.
.IP \[bu]
When the job status is finally known, if it was not known when it terminated.
Such lines are followed by the content of stderr if any.
.LP
During execution, if launched from a terminal, \fBlmake\fR also generates a progression status in the title bar.
This progression status line contains :
.IP \[bu]
The number of executed jobs (split between useful, rerun and hit)
.IP \[bu]
The number of queued jobs, i.e. jobs that are waiting for resources to be run
.IP \[bu]
The number of waiting jobs, i.e. jbos that are waiting for dependencies to be run
.LP
At the end of the execution, if the asked targets are not successfully generated, a summary is generated reminding the errors (with a configurable max, 20 by default)
and the stderr of the first error. Intermediate targets are deemed of secondary importance.
.LP
Before processing arguments, \fBlmake\fR prepends the content of \fB$LMAKE_ARGS\fR, separated by spaces.
This eases the management of user preferences.
For example, a user may like to systematically pass the \fF-a\fR and @\fFt\fR options, in which case they can set \fBLMAKE_ARGS\fR=\fB-a -t\fR.


.SH OPTIONS

.LP
Options common to all tools of the \fBopen-lmake\fI set of utilities :
.TP
\fB-h\fR, \fB--help\fR
Print a short help and exit.
.TP
\fB-J\fR, \fB--job\fR
Passed arguments are interpreted as job names rather than as file names.
Job names are the names that appear, for example, on start and done lines when \fBlmake\fR executes a job.
.TP
\fB-R\fR \fIrule\fR, \fB--rule\fR=\fIrule\fR
When the \fB-J\fR option is used, this options allows the specification of a rule, given by its name.
This is necessary when the job name is ambiguous as several rules may lead to the same job name.
.TP
\fB-q\fR, \fB--quiet\fR
Do not generate user oriented messages.
Strictly generate what is asked.
This is practical if output is meant for automatic processing.
.TP
\fB-S\fR, \fB--sync\fR
Ensure server is launched (i.e. do not connect to an existing server) and wait for its termination.
.TP
\fB-V\fR \fImode\fR, \fB--video\fR=\fImode\fR
Explicitly ask for a video mode instead of interrogating connected terminal.
If mode starts with \fBn\fR or \fBN\fR, normal video (black on white) is assumed.
If it starts with \fBr\fR or \fBR\fR, reverse video (white on black) is assumed.
Else output is not colorized.
video mode has an impact on generated colors as nice looking colors are not the same in each case.

.LP
Options specific to \fBlmake\fR :

.TP
\fB-a\fR, \fB--archive\fR
Ensure all intermediate files are up to date, in addition to the asked targets.
This is useful for example if you want to archive a fully built repository.
.TP
\fB-b\fR \fIvalue\fR, \fB--backend\fR=\fIvalue\fR
Pass value to backend.
This is used for example to pass a partition or specificities to the slurm backend for a particular command.
Note that backend only impacts resources and scheduling, not the content of the targets, so specifying such an option does not hurt repeatability.
.TP
\fB-e\fR, \fB--forget-old-errors\fR
Assume encountered errors (before this command) are transicent.
Contrarily to the \fBlforget -e\fR command, this only concerns this execution, not subsequent ones.
.TP
\fB-j\fR jobs, \fB--jobs\fR=\fIjobs\fR
When this option is used, \fBlmake\fR will limit the overall number of simultaneous jobs to \fIjobs\fR per backend.
If several \fBlmake\fR commands run simultaneously, a job cannot be launched on behalf of a given command if the number of running jobs is not less than its associated \fIjobs\fR.
.TP
\fB-o\fR, \fB--live-out\fR
Normally, \fFlmake\fR does not output the stdout of jobs (such stdout is accessible with the \fBlshow -o\fR command).
However, sometimes it is practical to have the output while jobs are running.
Generating such output for all jobs would produce an intermixed flow of characters of all jobs running in parallel making such an output unreadable.
When this option is used, only the jobs directly producing the asked targets have their output generated on the output of \fBlmake\fR.
Because most of the time there is a single target, this ensures that there is a single job generating its output, avoiding the intermixing problem.
.TP
\fB-r\fR \fIcount\fR, \fB--retry-on-error\fR=\fIcount\fR
Ask \fBlmake\fR to retry jobs in case of error.
The number of retries is the max between count and the n_retries attribute of the rule.
This is useful for unattended execution (e.g. nightly regressions) when system reliability is not enough to guarantee correct execution at the desired level.
.TP
\fB-l\fR, \fB--local\fR
With this option, jobs are launched locally (i.e. using the \fIlocal\fR backend) instead of the backend mentioned in the rule.
Note that if 2 \fBlmake\fR commands with different values for this option are running simultaneously, in case a job is necessary for both, it may be launched locally or remotely.
The originally targetted backend is in charge of mapping required resources mentioned in the rule to local resources understandable by the local backend.
.TP
\fB-s\fR, \fB--source-ok\fR
Normally, \fBlmake\fR refuses to launch a job that may overwrite a source.
With this option, the user instructs \fBlmake\fR that this is allowed.
.TP
\fB-t\fR, \fB--keep-tmp\fR
Normally, \fBlmake\fR washes the temporary directory allocated to a job at the end of job execution.
With this option, the user instructs \fBlmake\fR to keep the temporary directories, as if the \fIkeep_tmp\fR attribute was set for all rules.
The kept temporary directory can be retreived with \fBlshow -i\fR.
.TP
\fB-v\fR, \fB--verbose\fR
Enable the generation of some execution information from backend.
This is not done systematicly as this may incur a performance hit.
These information are available by using \fBlshow -i\fR.

.SH "EXIT STATUS"
.LP
\fBlmake\fR exits with a status of zero if the asked targets could be built or were already up-to-date.
Else it exits with a non-zero status.

.SH OUTPUT
.LP
While \fblmake\fR runs, it outputs a log.
This log is also recorded in \fILMAKE/outputs/<start date>\fR with the following differences :
.IP \[bu]
It is not colored.
.IP \[bu]
Reported files are relative to the root of the repository, not to the current working directory where the \fBlmake\fR command has been launched.
.LP
The log contains a line, possibly followed by attached information when the following events occur :
.IP \[bu]
A job is started, if the job duration is longer than the \fIstart_delay\fR attribute of the rule.
.IP \[bu]
A job is completed.
.IP \[bu]
A job status is known, while it was not known when it completed.
.IP \[bu]
A source file has been seen as modified.
.IP \[bu]
A frozen file or a target of a frozen job is needed.
.LP
Once the build process is complete, a summary is generated with :
.IP \[bu]
The frozen files and jobs that we necessary to carry out the build.
.IP \[bu]
The jobs in error (the first of them is accompanied with its stderr).

.SH ENVIRONMENT
.LP
The content of \fB$LMAKE_ARGS\fR is prepended to command line arguments.
.LP
Unless explicitely asked in \fILmakefile.py\fR, the environment is mostly ignored when \fBlmake\fR is run, i.e. it is not passed to the jobs.
The goal is to improve repeatability by protecting jobs from the variability environment variables may cause.
In particular :
.IP \[bu]
\fB$HOME\fR is redirected to the root of the repository.
This protects the job from all specificities stored in \fI.xxxrc\fR files in the home directory.
.IP \[bu]
\fB$LMAKE_ARGS\fR, although used by \fBlmake\fR, is not passed to jobs.
.IP \[bu]
\fB$PATH\fR is reset to the default path for the system, plus the \fIopen-lmake\fR bin directory.
.IP \[bu]
\fB$PYTHONPATH\fR is set to the \fIopen-lmake\fR lib directory.
.IP \[bu]
\fB$TMPDIR\fR is redirected to an isolated, empty directory which is cleaned up at the end of each job execution.
This way, the job can freely use this directory and need not take care of clean-up.
.LP
Moreover, a few variables are set during job execution :
.IP \[bu]
\fB$JOB_ID\fR is set to an integer specific of a job.
It does not change between executions, but may be different in different repository, even if strictly identical.
.IP \[bu]
\fB$SMALL_ID\fR is set to a as small as possible integer such that a different value is set for jobs running concurrently.
.IP \[bu]
\fB$SEQUENCE_ID\fR is set to a different value each time a job is run, they are never recycled.

.SH FILES
.LP
The file \fILmakefile.py\fR is searched in the current directory and in parent directories.
If a single one is found, this determines the root of the repository.
If several are found, the existence of an \fILMAKE\fR directory is checked.
If a single one is found, this determines the root of the repository.
In other cases, \fBlmake\fR will not start.

.SH "SEE ALSO"
.LP
\fBautodep\fR(1),
\fBfind_cc_ld_library_path\fR(1),
\fBlcheck_deps\fR(1),
\fBldebug\fR(1),
\fBldecode\fR(1),
\fBlencode\fR(1),
\fBlforget\fR(1),
\fBlmark\fR(1),
\fBlrepair\fR(1),
\fBlshow\fR(1),
\fBltarget\fR(1),
\fBxxhsum\fR(1).
.LP
The python module \fBlmake\fR.

.SH "COPYRIGHT"
Copyright \(co 2023-2024, Doliam.
This file is part of
.IR "open-lmake" .
.LP

debian/changelog
open-lmake (24.09.1-1) UNRELEASED; urgency=low

  * support Linux namespaces

 -- Cesar Douady <cesar.douady@doliam.fr>  Tue, 03 Sep 2024 15:56:35 +0200


lib/lmake.py
# This file is part of the lmake distribution (git@github.com:cesar-douady/open-lmake.git)
# Copyright (c) 2023 Doliam
# This program is free software: you can redistribute/modify under the terms of the GPL-v3 (https://www.gnu.org/licenses/gpl-3.0.html).
# This program is distributed WITHOUT ANY WARRANTY, without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

# there are 2 entirely different modules here :
# - support module to read lmakefile
# - support module during job execution
# but for user convenience, we want to name both lmake

native_version = (1,0)

import sys     as _sys
import os.path as _osp

from clmake import *                                                           # if not in an lmake repo, root_dir is not set to current dir

Kilo = 1_000                           # convenient constants
Mega = 1_000_000                       # .
Giga = 1_000_000_000                   # .
inf  = float('inf')                    # use the repr value as variable name

def version(major,minor) :
	'''check version'''
	if major!=native_version[0] or minor<native_version[1] : raise RuntimeError(f'required version {(major,minor)} is incompatible with native version {native_version}')

class pdict(dict) :
	def __getattr__(self,attr) :
		try             : return self[attr]
		except KeyError : raise AttributeError(attr)
	def __setattr__(self,attr,val) :
		try             : self[attr] = val
		except KeyError : raise AttributeError(attr)
	def __delattr__(self,attr) :
		try             : del self[attr]
		except KeyError : raise AttributeError(attr)

if getattr(_sys,'reading_makefiles',False) :
	#
	# read lmakefile
	#

	import os         as _os
	import pwd        as _pwd
	import signal     as _signal
	import subprocess as _sp

	_std_path = _sp.run( 'echo $PATH' , shell=True , check=True , stdout=_sp.PIPE , env={} , universal_newlines=True ).stdout.strip()
	def _get_std(cmd) :
		for d in _std_path.split(':') :
			abs_cmd = _osp.join(d,cmd)
			if _os.access(abs_cmd,_os.X_OK) : return abs_cmd
		raise FileNotFoundError(f'cannot find {cmd} in standard path {_std_path}')
	_lmake_dir    = _osp.dirname(_osp.dirname(__file__))
	_python       = _sys.executable
	_physical_mem = _os.sysconf('SC_PHYS_PAGES') * _os.sysconf('SC_PAGE_SIZE')

	# Lmakefile must define the following variables in module lmake :
	# - config  : the server configuration
	# - rules   : no default, must be set by user as the list of classes representing rules (use Rule & AntiRule base classes to help)
	# - sources : defaults to files listed in Manifest or by searching git

	exclude_deps = (
		'/bin/'
	,	'/boot/'
	,	'/dev/'
	,	'/etc/'
	,	'/lib/'
	,	'/proc/'
	,	'/run/'
	,	'/usr/'
	,	'/var/'
	)

	#
	# config
	#

	local_admin_dir  = ''                                  # directory in which to store data that are private to the server (not accessed remote executing hosts) (default is to use LMAKE dir)
	remote_admin_dir = ''                                  # directory in which to store tmp data during remote job execution (not used when keep_tmp is enforced) (default is to use LMAKE dir)

	config = pdict(
		hash_algo       = 'Xxh'                            # algorithm to use to compute checksums on files, one of 'Xxh' or 'Md5'
	,	heartbeat       = 60                               # delay between heartbeat checks of running jobs
	,	interface       = ''                               # network interface that connect to remote hosts executing jobs
	,	link_support    = 'Full'                           # symlinks are supported. Other values are 'None' (no symlink support) or 'File' (symlink to file only support)
	,	max_dep_depth   = 1000                             # used to detect infinite recursions and loops
#	,	max_error_lines = 30                               # used to limit the number of error lines when not reasonably limited otherwise
	,	network_delay   = 3                                # delay between job completed and server aware of it. Too low, there may be spurious lost jobs. Too high, tool reactivity may rarely suffer.
	,	trace_size      = 100*Mega                         # size of trace
#	,	path_max        = 400                              # max path length, but a smaller value makes debugging easier (by default, not activated)
	,	sub_prio_boost  = 1                                # increment to add to rules defined in sub-repository (multiplied by directory depth of sub-repository) to boost local rules
	,	console = pdict(                                   # tailor output lines
			date_precision = None                          # number of second decimals in the timestamp field
		,	host_length    = None                          # length of the host field (lines will be misaligned if a host is longer)
		,	has_exec_time  = True                          # if True, output the exec_time field
		)
	,	backends = pdict(                                  # PER_BACKEND : provide a default configuration for each backend
			local = pdict(                                 # besides margin, other entries mention the total availability of resources
				margin  = 0                                # compulsery in all backends, the margin to take (in seconds) to consider a newly discovered dep as reliable
		#	,	address = socket.getfqdn()                 # address at which lmake can be contacted from jobs launched by this backend, can be :
				#                                            - ''                     : loop-back address (127.0.0.1)
				#                                            - standard dot notation  : for example '192.168.0.1'
				#                                            - network interface name : the address of the host on this interface (as shown by ifconfig)
				#                                            - a host name            : the address of the host as found in networkd database (as shown by ping)
				#                                            default is loopback for local backend and hostname for the others
			,	cpu     = len(_os.sched_getaffinity(0))    # total number of cpus available for the process, and hence for all jobs launched locally
			,	mem     = _physical_mem//Mega              # total available memory (in MB)
			)
		)
	,	caches = pdict(                                    # PER_CACHE : provide an explanation for each cache method
	#		dir = pdict(                                   # when rule specifies cache = 'dir' , this cache is selected
	#			tag      'dir'                             # specify the caching method, must be one of the supported method
	#		,	repo   = root_dir                          # an id that identifies the repository, no more than one entry is stored in the cache for a given job and tag
	#		,	dir    = '/cache_dir'                      # the directory in which cached results are stored
	#		,	size   = 10*Giga                           # the overall size of this cache
	#		,	group  = os.getgroups()[0]                 # the group used to write to the cache. If user does not belong to this group, read-only access is still possible
	#		)
		)
	,	colors = pdict(
			#               normal video    reverse video
			hidden_note = ( (192,192,192) , ( 96, 96, 96) ) # gray
		,	hidden_ok   = ( (176,208,176) , ( 80,112, 80) ) # greenish gray
		,	note        = ( (  0,  0,255) , ( 64,160,255) ) # blue
		,	ok          = ( (  0,128,  0) , (128,255,128) ) # green
		,	warning     = ( (155,  0,255) , (255,  0,255) ) # magenta
		,	err         = ( (155,  0,  0) , (255,  0,  0) ) # red
		)
	)

	#
	# rules
	#
	rules = []                                             # list of rules that must be filled in by user code

	class _RuleBase :
		def __init_subclass__(cls) :
			rules.append(cls)
		# combined attributes are aggregated, combine is itself combined but not listed here as it is dealt with separately
		# for dict, a value of None   discards the entry, sequences are mapped to dict with value True and non sequence is like a singleton
		# for set , a key   of -<key> discards <key>    , sequences are mapped to set                  and non sequence is like a singleton
		# for list & tuple, entries are concatenated    , sequences are mapped to list/tuple           and non sequence is like a singleton
		combine = {
			'stems'
		,	'targets'     , 'post_targets'
		,	'deps'
		,	'environ_cmd' , 'environ_rsrcs' , 'environ_ancillary'
		,	'resources'
		}
	#	name                                               # must be specific for each rule, defaults to class name
	#	job_name                                           # defaults to first target
		stems = {                                          # defines stems as regexprs for use in targets & deps, e.g. {'File':'.*'}
			'__dir__'  : r'.+/|'
		,	'__file__' : r'.+'
		,	'__base__' : r'[^/]+'
		}
		targets      = {}                                  # patterns used to trigger rule execution, refers to stems above through {} notation, e.g. {'OBJ':'{File}.o'}
		#                                                  # in case of multiple matches, first matching entry is considered, others are ignored
		#                                                  # targets may have flags (use - to reset), e.g. {'TMP' : ('{File}.tmp/{Tmp*}','Phony','-Match','-Crc') }, flags may be :
		#                                                  #   flag         | default | description
		#                                                  #   -------------+---------+--------------------------------------------------------------------------------------------
		#                                                  #   Crc          | x       | compute CRC on target
		#                                                  #   Dep          | x       | make it a Dep if accessed read-only
		#                                                  #   Essential    | x       | show in graphic flow
		#                                                  #   Incremmental |         | file is not unlinked before job execution and can be read before written to or unlinked
		#                                                  #   ManualOk     |         | no error if file is modified outside lmake
		#                                                  #   Match        | x       | can match to select rule to rebuild this target
		#                                                  #   Phony        |         | file may not be generated or can be unlinked
		#                                                  #   SourceOk     |         | no error if file is a source that will be overwritten (e.g. by asking another target)
		#                                                  #   Star         |         | force target to be a star if no star stem appears (next rule will be tried if not generate)
		#                                                  #   Stat         | x       | consider stat-like accesses as reads, else ignore them
		#                                                  #   Warning      | x       | warn if unlinked by lmake before job execution and file was generated by another rule
		#                                                  #   Write        | x       | file may be written to
		post_targets = {}                                  # same as targets, except that they are processed after targets and in reverse order
	#	target                                             # syntactic sugar for targets      = {'<stdout>':<value>} (except that it is allowed)
	#	post_target                                        # syntactic sugar for post_targets = {'<stdout>':<value>} (except that it is allowed)

	class Rule(_RuleBase) :
		allow_stderr = False                               # if set, writing to stderr is not an error but a warning
		__anti__     = False                               # plain Rule
		auto_mkdir   = False                               # auto mkdir directory in case of chdir
		backend      = 'local'                             # may be set anywhere in the inheritance hierarchy if execution must be remote
		chroot       = ''                                  # chroot directory to execute cmd (if empty, chroot cmd is not done)
		cache        = None                                # cache used to store results for this rule. None means no caching
	#	cmd                                                # runnable if set anywhere in the inheritance hierarchy (as shell str or python function), chained if several definitions
	#	cwd                                                # cwd in which to run cmd. targets/deps are relative to it unless they start with /, in which case it means top root dir
	#                                                      # defaults to the nearest root dir of the module in which the rule is defined
		deps         = {}                                  # patterns used to express explicit depencies, refers to stems through f-string notation, e.g. {'SRC':'{File}.c'}
		#                                                  # deps may have flags (use - to reset), e.g. {'TOOL':('tool','-Essential')}, flags may be :
		#                                                  #   flag         | default | description
		#                                                  #   -------------+---------+--------------------------------------------------------------------------------------------
		#                                                  #   Essential    | x       | show in graphic flow
		#                                                  #   Required     | x       | requires that dep is buildable
	#	dep                                                # syntactic sugar for deps = {'<stdin>':<value>} (except that it is allowed)
		ete          = 0                                   # Estimated Time Enroute, initial guess for job exec time (in s)
		force        = False                               # if set, jobs are never up-to-date, they are rebuilt every time they are needed
		ignore_stat  = False                               # if set, stat-like syscalls do not, by themselves, trigger dependencies (but link_support is still ensured at required level)
		keep_tmp     = False                               # keep tmp dir after job execution
		kill_sigs    = (_signal.SIGKILL,)                  # signals to use to kill jobs (send them in turn, 1s apart, until job dies, 0's may be used to set a larger delay between 2 trials)
		local_marker = '$CWD'                              # a marker recognized in environ_* attributes and replaced by the cwd of the rule to allow cache effenciency
		n_tokens     = 1                                   # number of jobs likely to run in parallel for this rule (used for ETA estimation)
		prio         = 0                                   # in case of ambiguity, rules are selected with highest prio first
		python       = (_python,)                          # python used for callable cmd
		shell        = (_get_std('bash'),)                 # shell  used for str      cmd (_sh is usually /bin/sh which may test for dir existence before chdir, which defeats auto_mkdir)
	#	start_delay  = 1                                   # delay before sending a start message if job is not done by then
	#	stderr_len   = 20                                  # maximum number of stderr lines shown in output (full content is accessible with lshow -e)
		timeout      = None                                # timeout allocated to job execution (in s), must be None or an int
		job_tokens   = 1                                   # number of tokens taken by a job, follow the same syntax as deps (used for ETA estimation)
		if has_ld_audit : autodep = 'ld_audit'             # may be set anywhere in the inheritance hierarchy if autodep uses an alternate method : none, ptrace, ld_audit, ld_preload
		else            : autodep = 'ld_preload'           # .
		resources = {                                      # used in conjunction with backend to inform it of the necessary resources to execute the job, same syntax as deps
			'cpu' : 1                                      # number of cpu's to allocate to jobs
		,	'mem' : config.backends.local.mem//config.backends.local.cpu # memory to allocate to jobs
		}                                                  # follow the same syntax as deps
		environ_cmd = pdict(                               # job execution environment, handled as part of cmd (trigger rebuild upon modification)
			HOME       = '$CWD'                            # favor repeatability by hiding use home dir some tools use at start up time
		,	PATH       = ':'.join(( _lmake_dir+'/bin' , _std_path ))
		,	PYTHONPATH = ':'.join(( _lmake_dir+'/lib' ,           ))
		)
		environ_resources = pdict()                        # job execution environment, handled as resources (trigger rebuild upon modification for jobs in error)
		environ_ancillary = pdict(                         # job execution environment, does not trigger rebuild upon modification
			UID  = str(_os.getuid())                       # this may be necessary by some tools and usually does not lead to user specific configuration
		,	USER = _pwd.getpwuid(_os.getuid()).pw_name     # .
		)

	class AntiRule(_RuleBase) :
		__anti__ = True                                    # AntiRule's are not executed, but defined at high enough prio, prevent other rules from being selected
		prio     = float('inf')                            # default to high prio as the goal of AntiRule's is to hide other rules

	class GitRule(Rule) :
		'base rule that ignores read accesses (and forbid writes) to git administrative files'
		post_targets = { '__GIT__' : ( '{__dir__*}.git/{__file__*}' , '-Dep','Incremental','-Match','-Write' ) }

	class HomelessRule(Rule) :
		'base rule to redirect the HOME environment variable to TMPDIR'
		def cmd() :
			import os
			os.environ['HOME'] = os.environ['TMPDIR']
		cmd.shell = 'export HOME=$TMPDIR'                                      # defining a function with a shell attribute is the way to have both python & shell pre-commands

	class DirtyRule(Rule) :
		post_targets = { '__NO_MATCH__' : ('{*:.*}','-crc','-essential','incremental','-match','-warning') }

	class PyRule(Rule) :
		'base rule that handle pyc creation when importing modules in Python'
		def __init_subclass__(cls) :
			super().__init_subclass__()
			if cls.ignore_stat and '-B' not in cls.python[1:] :
				raise AttributeError('if the ignore_stat attribute is set on one rule, python must run with the -B flag set on all rules')
		# ignore pyc files, Python takes care of them
		post_targets = { '__PYC__' : ( r'{__dir__*}__pycache__/{__base__*}.{__cache_tag__*:\w+-\d+}.pyc' , '-Dep','Incremental','ManualOk','-Match','-Crc' ) }

	class DynamicPyRule(PyRule) :                                              # base rule that handle import of generated modules in Python
		virtual = True
		def cmd() :                                                            # this will be executed before cmd() of concrete subclasses as cmd() are chained in case of inheritance
			fix_imports()

	class RustRule(Rule) :
		'base rule for use by any code written in Rust (including cargo and rustc that are written in rust)'
		autodep = 'ld_preload'

	#
	# sources
	#

#sources = ()      # this variable may be set by Lmakefile.py to provide the list of sources, if it is not, auto_sources() is called to fill it

	def manifest_sources(manifest='Manifest',**kwds) :
		'''
			read manifest, filtering out comments :
			- comments start with #
			- they must be separated from file with spaces
			- files must start and end with non-space and cannot start with #
			- note that files can contain spaces (except first and last character) and # as long as they are not preceded by spaces
			kwds are ignored which simplifies the usage of auto_sources
		'''
		import re
		line_re = re.compile(r'\s*(?P<file>[^#\s](.*\S)?)?\s+(#.*)?\n?')
		return [ f for f in ( line_re.fullmatch(l).group('file') for l in open(manifest) ) if f ]

	def git_sources( recurse=True , ignore_missing_submodules=False , **kwds ) :
		'''
			gather git controled files.
			recurse to sub-modules if recurse is True
			kwds are ignored which simplifies the usage of auto_sources
		'''
		git_top = _osp.isdir('.git')

		git_cmd = ['git','ls-files']
		if recurse : git_cmd.append('--recurse-submodules')
		try :
			srcs = _sp.run( git_cmd , check=True , stdout=_sp.PIPE , stderr=_sp.DEVNULL , universal_newlines=True ).stdout.splitlines()
		except :
			if recurse : return git_sources(False)
			else       : raise
		if not srcs : raise FileNotFoundError(f'cannot find Lmakefile.py in git files')
		if git_top : srcs += [ _osp.join('.git',f) for f in ('HEAD','config','index') ] # not listed by git, but actually sources if .git is in repo as they are read by git ls-files

		if recurse :
			git_cmd    = ('git','submodule','--quiet','foreach','--recursive','echo $displaypath')
			submodules = _sp.run( git_cmd , check=True , stdout=_sp.PIPE , stderr=_sp.DEVNULL , universal_newlines=True ).stdout.splitlines()
			for submodule in submodules :
				if submodule.startswith('../') : continue
				sub_git = _osp.join(submodule,'.git')
				# XXX : fix using 'git submodule status' to detect and handle cases where submodules are not loaded but this creates more dependencies
				if not _osp.exists(sub_git) and not ignore_missing_submodules : raise FileNotFoundError(f'cannot find git submodule {submodule}')
				git_dir      = open(sub_git).readline().split()[1]
				sub_git_dir  = _osp.normpath(_osp.join(_osp.dirname(sub_git),git_dir))
				srcs        += ( sub_git , *( _osp.join(sub_git_dir,f) for f in ('HEAD','config','index') ) ) # not listed by git, but actually a source as it is read by git ls-files
		return srcs

	def auto_sources(**kwds) :
		'''
			auto-determine used source control
			kwds can be passed to underlying source control
		'''
		for ctl in ( manifest_sources , git_sources ) :
			try    : return ctl(**kwds)
			except : pass
		raise FileNotFoundError(f'no source control found')

#
# job execution, available in rules to have full access when writing rule cmd's
#

module_suffixes = ('.so','.py','/__init__.py')                                 # can be tailored to suit application needs, the lesser the better (less spurious dependencies)

import importlib.machinery as _machinery

_std_suffixes = _machinery.all_suffixes()+['/__init__.py']                     # account for packages, not included in all_suffixes()
_local_admin  = 'LMAKE/'
_global_admin = root_dir+'/'+_local_admin

def _maybe_local(file) :
	'fast check for local files, avoiding full absolute path generation'
	return not file or file[0]!='/' or file.startswith(root_dir)

def fix_imports() :
	'''fixes imports so as to be sure all files needed to do an import is correctly reported (not merely those which exist)'''
	class Depend :
		@staticmethod
		def find_spec(module_name,path,target=None) :
			if path==None : path = _sys.path
			tail = module_name.rsplit('.',1)[-1]
			for dir in path :
				if dir : dir += '/'
				base = dir+tail
				if _maybe_local(base) :
					for suffix in module_suffixes :
						file = base+suffix
						depend(file,required=False,essential=False)
						if _osp.exists(file) : return
				else :
					for suffix in _std_suffixes :
						if _osp.exists(base+suffix) : return

	# give priority to system so as to avoid too numerous dependencies
	_sys.path = (
		[ d for d in _sys.path if not _maybe_local(d+'/') ]
	+	[ d for d in _sys.path if     _maybe_local(d+'/') ]
	)

	# put dependency checker before the first path based finder
	for i in range(len(_sys.meta_path)) :
		if _sys.meta_path[i]==_machinery.PathFinder :
			_sys.meta_path.insert(i,Depend)
			break
	else :
		_sys.meta_path.append(Depend)
